ddtrace/internal/processor/trace.py

```diff
@@ -0,0 +1,127 @@
+import abc
+from collections import defaultdict
+import threading
+from typing import DefaultDict
+from typing import Iterable
+from typing import List
+from typing import Optional
+
+import attr
+import six
+
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.processor import SpanProcessor
+from ddtrace.internal.writer import TraceWriter
+from ddtrace.span import Span
+
+
+log = get_logger(__name__)
+
+
+@attr.s
+class TraceProcessor(six.with_metaclass(abc.ABCMeta)):
+    def __attrs_post_init__(self):
+        # type () -> None
+        """Default post initializer which logs the representation of the
+        TraceProcessor at the ``logging.DEBUG`` level.
+
+        The representation can be modified with the ``repr`` argument to the
+        attrs attribute::
+
+            @attr.s
+            class MyTraceProcessor(TraceProcessor):
+                field_to_include = attr.ib(repr=True)
+                field_to_exclude = attr.ib(repr=False)
+        """
+        log.debug("initialized trace processor %r", self)
+
+    @abc.abstractmethod
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        """Processes a trace.
+
+        None can be returned to prevent the trace from being exported.
+        """
+        pass
+
+
+@attr.s
+class TraceSamplingProcessor(TraceProcessor):
+    def process_trace(self, trace):
+        # type: (List[Span]) -> Optional[List[Span]]
+        sampled_spans = [s for s in trace if s.sampled]
+        if len(sampled_spans) == 0:
+            log.info("dropping trace, %d spans unsampled", len(trace))
+            return None
+
+        log.info("trace %d sampled (%d/%d spans sampled)", trace[0].trace_id, len(sampled_spans), len(trace))
+        return trace
+
+
+@attr.s
+class SpansToTraceProcessor(SpanProcessor):
+    """Processor that aggregates spans together by trace_id and submits the
+    finalized trace if the trace is assumed to be complete[0] or if a partial
+    flushing threshold has been met[1].
+
+    [0] A trace is assumed to be complete if all the spans that have been
+    created with the trace_id have finished. When this condition is met the
+    completed spans are returned.
+
+    [1] A minimum threshold of spans can be specified with the
+    ``partial_flush_min_spans`` argument. When this threshold of spans has been
+    finished the finished spans of the trace are returned.
+    """
+
+    @attr.s
+    class _Trace(object):
+        spans = attr.ib(default=attr.Factory(list))  # type: List[Span]
+        num_finished = attr.ib(type=int, default=0)  # type: int
+
+    _partial_flush_enabled = attr.ib(type=bool)
+    _partial_flush_min_spans = attr.ib(type=int)
+    _trace_processors = attr.ib(type=Iterable[TraceProcessor])
+    _writer = attr.ib(type=TraceWriter)
```

Comment:
By passing in a writer, I feel like we are overloading a `Processor` by asking it to do more than it was designed for. I would expect a processor to just process, and leave the task of writing the result to other dedicated structures. WDYT?

Reply:
Agreed and I cover this a bit in the PR description (at the end). The problem is that the data that the writer consumes is coupled to the implementation of the processor.

It looks bad right now because the writer outwardly appears to be a modular component. Internally it's quite coupled to the data format (lists of spans) and wouldn't be reusable for another kind of span format.

We tried in #2210 to have it so the writer was independent but I don't think it makes sense if there are multiple processors, writers, different formats, etc.

Significant pieces of the writer could probably be pulled out into the processor: serialization, payload creation and then sending. I'm not 100% about this yet so let me know what you think! 🙂

Reply:
I can see how the writer is essentially coupled with the data format but I feel like the tracer should remain in control of the writer, while the processors just enqueue their result somewhere, ready to be written, and ask the tracer to write via the on_finish callback. Of course this is essentially not too different from writing directly, but breaking this coupling might give benefits on the long run?

Reply:
Summary from offline discussion: Gab makes a good point about separating the http/sending logic in the writer (right now serialization logic is coupled with the sending of traces) which I think we all agree with.

Another complication might be with multiple writers (eg. AgentWriter, ConsoleWriter; partial flush to the AgentWriter but want to send complete traces to the ConsoleWriter). This kind of scenario seems to be incompatible with having modular writers managed by the tracer.  

I'm going to update the PR description with the results of the discussion.

Ultimately though we are ok with the approach proposed in the PR in the interim in order to get this change merged as it's been blocking the context work. It's easy enough to change/refactor the writer usage in the future (only one reference and function call).